### PR TITLE
refactor & add better retry logic to NewTestAgent

### LINKED ISSUE #6361: Add 3x retry to agent.NewTestAgent, model based on api.makeClientWithConfig.
Existing internal retry must be scaled down to 2 seconds.

Original card created by @freddygv for flaky test pile.

### PR DESCRIPTION
**Pull request by adams-sarah**
**Description updated 8/21**

**Note to reviewers**: the only real logic changes are in `agent/testagent.go`.

Code changes:
1. Add retry logic (3x retry) to within `NewTestAgent` to match the new retry logic already in `api.makeClientWithConfig`.
2. Change `agent.Start()` to return an error instead of calling `t.Fatal` so that the function can be retried.
3. Change the internal retry logic within `agent.Start()` to again match the retry logic of `api.makeClientWithConfig` (retry for 2 seconds with 25ms wait between tries).

Fixes #6361

### CODE REVIEW

File: agent/testagent.go [hunk 1]
```diff
@@ -184,18 +204,36 @@ func (a *TestAgent) Start(t *testing.T) *TestAgent {
...
+	deadline := time.Now().Add(timer.Timeout)
+
+	var retErr error
 	var out structs.IndexedNodes
-	retry.Run(t, func(r *retry.R) {
+	for !time.Now().After(deadline) {
 		if len(a.httpServers) == 0 {
-			r.Fatal(a.Name, "waiting for server")
+			retErr = fmt.Errorf("%s: waiting for server", a.Name)
+			continue // fail, try again
```
Comment (mkeeler) [inline agent/testagent.go]: All the `continue` statements here seem like they will avoid the `time.Sleep` at the end of the loop and burn cpu cycles. Should that sleep be moved to the beginning of the loop?

Alternatively you could wrap all of the checks in for loop like so and break instead of continue:

```go
for !time.Now().After(deadline) {
   for {
     if <check> {
       retErr = ...
       break
     }
   }
   
   time.Sleep(timer.Wait)
}
```

That way the sleep can be avoided if everything is already ready but we can still prevent the hot loop.

File: agent/testagent.go [hunk 2]
```diff
@@ -165,15 +185,15 @@ func (a *TestAgent) Start(t *testing.T) *TestAgent {
...
 			break
 		} else if i == 0 {
-			require.Fail("%s %s Error starting agent: %s", id, a.Name, err)
+			return fmt.Errorf("%s %s Error starting agent: %s", id, a.Name, err)
 		} else if a.ExpectConfigError {
-			// Panic the error since this can be caught if needed. Pretty gross way to
-			// detect errors but enough for now and this is a tiny edge case that I'd
-			// otherwise not have a way to test at all...
-			panic(err)
+			return fmt.Errorf("a.ExpectConfigError: %v", err)
```
Comment (freddygv) [inline agent/testagent.go]: I'm a bit confused by the rationale for why a `panic` was being called here. Is a test somewhere relying on recovering from this panic?

**Comment by mkeeler** [comment]
FYI: There are going to be some merge conflicts here after we merge release/1-6 into master for the final 1.6.0 release. All of the managed proxy tests were deleted (that are also modified in this PR). There are likely to be other conflicts too as many tests were added/udpated in that branch.

File: agent/testagent.go (hunk 2 above)
Comment (adams-sarah) [reply agent/testagent.go]: good question. when i deleted it, my intention was to just see if it broke any tests. i haven't finished debugging the tests yet, so i'll leave this here and get back to you.

File: agent/testagent.go (hunk 1 above)
Comment (adams-sarah) [reply agent/testagent.go]: oh duh! thanks matt, great catch. i'll add it to the loop header.

**Comment by adams-sarah** [comment]
@mkeeler 
> Update: I re-read the PR description and it says right there that we pass an error back so the caller can retry. Any pointers as to where this might actually be done?

Yes, totally. This is currently in `NewTestAgent` here: https://github.com/hashicorp/consul/pull/6363/files#diff-4248a5311e844c10537bd411253a4d82R97

> On the whole the retry logic in the test agent seem okay to me. One question I have is what advantage there is to not passing a *testing.T into the NewTestAgent function and allowing it to abort the test. Is there any place where we call NewTestAgent and do not immediately abort if there is a non-nil err?

You are totally right that the way I’ve implemented this change now, the `NewTestAgent` signature could have remained the same - we could have `t.Fatal`’d inside `NewTestAgent` upon failure instead of returning an error. Though, `t` would only ever be used this once.

In returning an error instead, we can let folks retry `NewTestAgent` and its helper-functions as much as they like, in whatever way they feel is best for their test.
Returning errors in test helper functions like this, instead of passing around `testing.T`, gives us more flexibility to leverage for retrying code upon failure/error, or otherwise dealing with errors more thoughtfully/gracefully when required.

TBH, when I was initially implementing this I wasn’t sure where the best place for this new retry logic would be — would it be within `NewTestAgent`? Or perhaps in the callers of `NewTestAgent`?
I think that as we massage out our flaky tests, this decision I’ve made to put the retry logic inside `NewTestAgent` will probably change, as will the retry logic itself I think (3x{2s{try}}). In massaging the function to return an error now, we create a lot of room to play with retry logic later — for you can’t retry a function once `t.Fatal` has been called!

Let me know what you think.

File: agent/testagent.go [hunk 3]
```diff
@@ -123,7 +134,9 @@ func (a *TestAgent) Start(t *testing.T) *TestAgent {
 		}
 		name = strings.Replace(name, "/", "_", -1)
 		d, err := ioutil.TempDir(TempDir, name)
-		require.NoError(err, fmt.Sprintf("Error creating data dir %s: %s", filepath.Join(TempDir, name), err))
+		if err != nil {
+			return fmt.Errorf("Error creating data dir %s: %s", filepath.Join(TempDir, name), err)
+		}
 		hclDataDir = `data_dir = "` + d + `"`
 	}
```
[the remark this answers is not in the document]
Comment (freddygv) [reply agent/testagent.go]: `Start()` needs the retry loop, the biggest issue these tests currently have is that the readiness retries in `waitForUp()` hang even with the default 7s retry. The idea in this PR is to be able to recover from that by shortening that readiness retry and retrying the whole thing 3x. The change has already be implemented int the api package and has led to big improvements.

re: the 10x inner loop, I do think that could be scrapped now. Instead relying on the outer 3x loop. (Or combining it with the 2sec `waitForUp`.

**Comment by mkeeler** [comment]
@adams-sarah I can see the flexibility here but are we utilizing that flexibility? I skimmed all the changed files but didn't see where we were doing anything besides just checking for the error and failing the test.

I think a good course of action would be to not change the signature and behavior of allowing the internal Start function to fail the test until we actually need it to be more flexible. 

A few changes I would make to this code are:

1. Create/Use an interface instead of a concrete `*testing.T`. This would be similar to how the `retry` code works. So if an external caller wants to retry the whole operations they could pass a `retry.R` to the setup. This would give about the same level of flexibility in terms of being able to retry without requiring that every place we create a test agent we also have to handle the same error. These are incredibly numerous and 99% of the time we just want to fail the test when the setup fails.

2. We should call `t.Helper()` in both constructors and the start method. The interface for number 1 above should require a `Helper()` method too. We would need

If we did those things then we don't need any of the modifications to the callers right now as `*testing.T` would already implement the interface.

**Comment by adams-sarah** [comment]
Thank you all for your feedback. I am going to remove the change to the `NewTestAgent` function signature. I will also update the code in `testagent.go` to reflect the changes some of you have requested there. I will ping here when this PR is ready for feedback again.

File: agent/testagent.go [hunk 4]
```diff
@@ -93,16 +90,28 @@ type TestAgent struct {
...
-// directories.
-func NewTestAgent(t *testing.T, name string, hcl string) *TestAgent {
-	a := &TestAgent{Name: name, HCL: hcl}
-	a.Start(t)
-	return a
+// configuration. It returns an error if the Agent could not be started.
+func NewTestAgent(name string, hcl string) (a *TestAgent, err error) {
+	a = &TestAgent{Name: name, HCL: hcl}
+
+	rt := retry.ThreeTimes()
```
[the remark this answers is not in the document]
Comment (adams-sarah) [reply agent/testagent.go]: ack

File: agent/testagent.go [hunk 5]
```diff
@@ -138,13 +151,18 @@ func (a *TestAgent) Start(t *testing.T) *TestAgent {
 
 		// write the keyring
 		if a.Key != "" {
-			writeKey := func(key, filename string) {
+			writeKey := func(key, filename string) error {
 				path := filepath.Join(a.Config.DataDir, filename)
-				err := initKeyring(path, key)
-				require.NoError(err, fmt.Sprintf("Error creating keyring %s: %s", path, err))
```
[the remark this answers is not in the document]
Comment (adams-sarah) [reply agent/testagent.go]: ack

File: agent/testagent.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (adams-sarah) [reply agent/testagent.go]: will remove the 10x inner loop

**Comment by adams-sarah** [comment]
PR ready for feedback again.

File: agent/testagent.go (hunk 2 above)
Comment (adams-sarah) [reply agent/testagent.go]: alright I figured out where this was being used. was going to refactor the code to fix the use of panic here, but just wrote a comment instead.

File: agent/testagent.go [hunk 6]
```diff
@@ -123,79 +131,109 @@ func (a *TestAgent) Start(t *testing.T) *TestAgent {
...
+
+		return fmt.Errorf("%s %s Error starting agent: %s", id, a.Name, err)
 	}
 
+	a.Agent = agent
+
 	// Start the anti-entropy syncer
 	a.Agent.StartSync()
 
+	err = a.waitForUp()
```
Comment (i0rek) [inline agent/testagent.go]: Nitpick: could be the same style as above `if err = a.waitForUp(); err != nil {`.

File: agent/testagent.go [hunk 7]
```diff
@@ -123,79 +131,109 @@ func (a *TestAgent) Start(t *testing.T) *TestAgent {
 		}
 		name = strings.Replace(name, "/", "_", -1)
 		d, err := ioutil.TempDir(TempDir, name)
```
[the remark this answers is not in the document]
Comment (rboyer) [reply agent/testagent.go]: Sidenote: the fact that it's commented out _by default_ is a fun nightmare to reverse. Almost no tests need to test resuming one server from the state of another instance of the server (a valid test, but not common). Rather than make the cleanup behavior opt-out, this change made it opt-in which means nothing really cleans up at all, and `/tmp` is full of litter which is fun.

I guess I was concerned mostly in the case where within one call of `TestAgent.Start()` you create a temp directory and then early return at any point before the eventual cleanup that happens when `Agent.Start()` fails. In that case you'd leak the temp directory.

File: agent/testagent.go [hunk 8]
```diff
@@ -123,79 +131,109 @@ func (a *TestAgent) Start(t *testing.T) *TestAgent {
...
 	}
 
+	a.Agent = agent
+
 	// Start the anti-entropy syncer
 	a.Agent.StartSync()
 
+	err = a.waitForUp()
+	if err != nil {
+		return err
```
[the remark this answers is not in the document]
Comment (rboyer) [reply agent/testagent.go]: Callers don't call shutdown if `TestAgent.Start` fails. They only invoke that if it succeeds.

File: agent/testagent.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (adams-sarah) [reply agent/testagent.go]: ok, i think i've got this one covered now. code push coming shortly.

File: agent/testagent.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (adams-sarah) [reply agent/testagent.go]: ack & done

File: agent/testagent.go [hunk 9]
```diff
@@ -93,15 +91,23 @@ type TestAgent struct {
...
+// configuration. It fails the test if the Agent could not be started.
+// The caller should call Shutdown() to stop the agent and remove
+// temporary directories.
 func NewTestAgent(t *testing.T, name string, hcl string) *TestAgent {
 	a := &TestAgent{Name: name, HCL: hcl}
-	a.Start(t)
+
+	retry.RunWith(retry.ThreeTimes(), t, func(r *retry.R) {
+		if err := a.Start(); err != nil {
+			a.Shutdown()
```
[the remark this answers is not in the document]
Comment (adams-sarah) [reply agent/testagent.go]: ack, done

**Comment by adams-sarah** [comment]
PTAnotherL! 
Note: I'm going to leave the merge conflicts for after y'all have accepted the code changes. Don't want to duplicate work if I have to rework this PR again.
